tket/src/Circuit/SubcircuitFinder.cpp

```diff
@@ -0,0 +1,285 @@
+  VertexSet verts;  // all vertices in the subcircuit
+  VertexSet preds;  // all predecessors of vertices in the subcircuit that are
+                    // not themselves in it
+  VertexSet succs;  // all successors of vertices in the subcircuit that are not
+                    // themselves in it
+} subcircuit_info_t;
+
+// Test whether the union of two connected convex subcircuits is connected.
+static bool union_is_connected(
+    const subcircuit_info_t &subcircuit_info0,
+    const subcircuit_info_t &subcircuit_info1) {
+  const VertexSet &verts0 = subcircuit_info0.verts;
+  const VertexSet &verts1 = subcircuit_info1.verts;
+  for (const Vertex &v : subcircuit_info0.succs) {
+    if (verts1.contains(v)) {
+      return true;
+    }
+  }
+  for (const Vertex &v : subcircuit_info1.succs) {
+    if (verts0.contains(v)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Return the union of two disjoint convex connected subcircuits, assuming that
+// the union is convex and connected.
+static subcircuit_info_t convex_union(
+    const subcircuit_info_t &subcircuit_info0,
+    const subcircuit_info_t &subcircuit_info1) {
+  const VertexSet &verts0 = subcircuit_info0.verts;
+  const VertexSet &preds0 = subcircuit_info0.preds;
+  const VertexSet &succs0 = subcircuit_info0.succs;
+  const VertexSet &verts1 = subcircuit_info1.verts;
+  const VertexSet &preds1 = subcircuit_info1.preds;
+  const VertexSet &succs1 = subcircuit_info1.succs;
+
+  // verts = verts0 ∪ verts1
+  VertexSet verts = verts0;
+  verts.insert(verts1.begin(), verts1.end());
+
+  // preds = (preds0 ∖ verts1) ∪ (preds1 ∖ verts0)
+  VertexVec p0mv1;
+  std::set_difference(
+      preds0.begin(), preds0.end(), verts1.begin(), verts1.end(),
+      std::inserter(p0mv1, p0mv1.begin()));
+  VertexVec p1mv0;
+  std::set_difference(
+      preds1.begin(), preds1.end(), verts0.begin(), verts0.end(),
+      std::inserter(p1mv0, p1mv0.begin()));
+  VertexSet preds{p0mv1.begin(), p0mv1.end()};
+  preds.insert(p1mv0.begin(), p1mv0.end());
+
+  // succs = (succs0 ∖ verts1) ∪ (succs1 ∖ verts0)
+  VertexVec s0mv1;
+  std::set_difference(
+      succs0.begin(), succs0.end(), verts1.begin(), verts1.end(),
+      std::inserter(s0mv1, s0mv1.begin()));
+  VertexVec s1mv0;
+  std::set_difference(
+      succs1.begin(), succs1.end(), verts0.begin(), verts0.end(),
+      std::inserter(s1mv0, s1mv0.begin()));
+  VertexSet succs{s0mv1.begin(), s0mv1.end()};
+  succs.insert(s1mv0.begin(), s1mv0.end());
+
+  return {verts, preds, succs};
+}
+
+// Helper class for finding connected convex subcircuits.
+class SubcircuitFinder {
+ public:
+  SubcircuitFinder(const Circuit *circ) : circ_(circ) {
+    // Compute the partial order on the DAG:
+    DAG TC;
+    boost::transitive_closure(circ_->dag, TC);
+    BGL_FORALL_VERTICES(v, circ_->dag, DAG) { order_relations_.insert({v, v}); }
+    BGL_FORALL_EDGES(e, TC, DAG) {
+      order_relations_.insert({boost::source(e, TC), boost::target(e, TC)});
```

Comment:
I hadn't thought of that. But I would assume not, since checking for the existence of an edge in a `boost::graph` is unlikely to be significantly faster than checking for membership of a pair in a `std::set` (which is logarithmic in the size of the set)?

Reply:
Never mind, I somehow thought that the complexity might be constant with `adjacency_list`, but some google search suggested that's only the case with `AdjacencyMatrix`.